File: param_management_cpp/examples/value_manager.cpp
// Copyright 2024 Simon Sagmeister
#include <iostream>

#include "param_management_cpp/param_value_manager.hpp"
class SoftwareModule
{
private:
  tam::pmg::ParamValueManager::SharedPtr param_manager =
    std::make_shared<tam::pmg::ParamValueManager>();
  void declare_parameters() const
  {
    // Declare the parameters
    param_manager->declare_parameter("a", 1.0, tam::pmg::ParameterType::DOUBLE, "Description of a");
    param_manager->declare_parameter("b", 2.0, tam::pmg::ParameterType::DOUBLE, "Description of b");
    param_manager->declare_parameter("c", 3.0, tam::pmg::ParameterType::DOUBLE, "Description of c");
  }

public:
  SoftwareModule()
  {
    // Declare the parameters here and initialize the values in the struct
    declare_parameters();
  }
  double step() const
  {
    // Be careful, getting a parameter directly from the manager each time is
    // comparably inefficient. It is better to get the value once and store it in a local variable.
    // For this, see the example "value_manager_struct.cpp"
    return param_manager->get_value("a").as_double() * param_manager->get_value("b").as_double() *
           param_manager->get_value("c").as_double();
  }
};
int main()
{
  auto module = SoftwareModule();
  std::cout << "Result: " << module.step() << std::endl;
  return 0;
}

File: param_management_py/examples/value_manager.py
# Copyright 2024 Simon Sagmeister
import param_management_py as pmg


class SoftwareModule:

    def __init__(self):
        self.param_manager = pmg.ParamValueManager()
        self.declare_parameters()

    def declare_parameters(self):
        # fmt: off
        self.param_manager.declare_parameter("a", 1.0, pmg.ParameterType.DOUBLE, "Description of a")  # noqa: E501
        self.param_manager.declare_parameter("b", 2.0, pmg.ParameterType.DOUBLE, "Description of b")  # noqa: E501
        self.param_manager.declare_parameter("c", 3.0, pmg.ParameterType.DOUBLE, "Description of c")  # noqa: E501
        # fmt: on

    def step(self):

        return (
            self.param_manager.get_value("a").as_double()
            * self.param_manager.get_value("b").as_double()
            * self.param_manager.get_value("c").as_double()
        )


if __name__ == "__main__":
    mod = SoftwareModule()
    print(f"Result: {mod.step()}")
